Task Priority Escalation APIs (#2685)
swiftlang/swift-evolution: proposals/NNNN-task-priority-escalation-apis.md

Another example of libraries which may want to reach for manual task priority escalation APIs are libraries which facilitate communication across process boundaries, and would like to react to priority escalation and propagate it to a different process. Relying on the built-in priority escalation mechanisms won't work, because they are necessarily in-process, so libraries like this need to be able to participate and be notified when priority escalation happens, and also be able to efficiently cause the escalation inside the other process.

## Proposed solution

In order to address the above use-cases, we propose to add a pair of APIs: to react to priority escalation happening within a block of code, and an API to _cause_ a priority escalation without resorting to trickery using creating new tasks whose only purpose is to escalate the priority of some other task:

```swift
let m: Mutex<Task<Void, Never>?> 

await withTaskPriorityEscalationHandler {
  await withCheckedContinuation { cc in 
    let t = Task { cc.resume() }
    m.withLock { $0 = t }
  } onPriorityEscalated: { newPriority in 
    let t = m.withLock { $0 }

>     Task.escalatePriority(t, to: newPriority)

Reviewer: I think you've got a type mismatch in this example, `t` is an `Optional`, given the type of `m`.

> await withTaskPriorityEscalationHandler {
>   await withCheckedContinuation { cc in 
>     let t = Task { cc.resume() }
>     m.withLock { $0 = t }
>   } onPriorityEscalated: { newPriority in 
>     let t = m.withLock { $0 }
>     Task.escalatePriority(t, to: newPriority)
>   }
> }

Reviewer 2: I think this is the right nesting
```suggestion
await withTaskPriorityEscalationHandler {
  await withCheckedContinuation { cc in 
    let t = Task { cc.resume() }
    m.withLock { $0 = t }
  }
} onPriorityEscalated: { newPriority in 
  let t = m.withLock { $0 }
  Task.escalatePriority(t, to: newPriority)
}
```

```

## Detailed design

We propose the addition of a task priority escalation handler, similar to task cancellation handlers already present in the concurrency library:

```swift
public func withTaskPriorityEscalationHandler<T, E>(
  operation: () async throws(E) -> T,
  onPriorityEscalated handler: @Sendable (TaskPriority) -> Void,
  isolation: isolated (any Actor)? = #isolation

> ) async throws(E) -> T

Author: That's not how stdlib shapes are (e.g. cancellation handler), we're staying consistent with what we have.
It's problematic if APIs randomly differ in order

# Task Priority Escalation APIs

* Proposal: [SE-NNNN](NNNN-filename.md)
* Authors: [Konrad 'ktoso' Malawski](https://github.com/ktoso)
* Review Manager: TBD

> * Status: **Implemented**

Reviewer 3: ```suggestion
* Status: **Active review (February 20...March 2, 2025)**
```

* Implementation: https://github.com/swiftlang/swift/pull/78625
* Review: ...

## Introduction

A large part of Swift Concurrency is its Structured Concurrency model, in which tasks automatically form parent-child relationships, and inherit certain traits from their parent task. For example, a task started from a medium priority task, also starts on the medium priority, and not only that – if the parent task gets awaited on from a higher priority task, the parent's as well as all of its child tasks' task priority will be escalated in order to avoid priority inversion problems.

This feature is automatic and works transparently for any structured task hierarchy. This proposal will discuss exposing user-facing APIs which can be used to participate in task priority escalation.

## Motivation

Generally developers can and should rely on the automatic task priority escalation happening transparently–at least for as long as all tasks necessary to escalate are created using structured concurrency primitives (task groups and `async let`). However, sometimes it is not possible to entirely avoid creating an unstructured task. 

> One such example is the async sequence [`merge`](https://github.com/apple/swift-async-algorithms/blob/main/Sources/AsyncAlgorithms/AsyncAlgorithms.docc/Guides/Merge.md) operation from the [swift-async-algorithms](https://github.com/apple/swift-async-algorithms/) project where the implementation is forced to create an unstructured task for iterating the upstream sequences, which must outlive downstream calls. These libraries would like to participate in task priority escalation to boost the priority of the upstream consuming task, however today lack the API to do so.

Reviewer 4: nit: use permalinks to the doc references (here & below) so it doesn't randomly break if it gets moved. (if you load the pages in GitHub and then press 'y' it will display the permalink format in the browser url bar).

Author: Yeah good idea: 

```suggestion
One such example is the async sequence [`merge`](https://github.com/apple/swift-async-algorithms/blob/4c3ea81f81f0a25d0470188459c6d4bf20cf2f97/Sources/AsyncAlgorithms/AsyncAlgorithms.docc/Guides/Merge.md) operation from the [swift-async-algorithms](https://github.com/apple/swift-async-algorithms/) project where the implementation is forced to create an unstructured task for iterating the upstream sequences, which must outlive downstream calls. These libraries would like to participate in task priority escalation to boost the priority of the upstream consuming task, however today lack the API to do so.
```

> One such example is the async sequence [`merge`](https://github.com/apple/swift-async-algorithms/blob/4c3ea81f81f0a25d0470188459c6d4bf20cf2f97/Sources/AsyncAlgorithms/AsyncAlgorithms.docc/Guides/Merge.md) operation from the [swift-async-algorithms](https://github.com/apple/swift-async-algorithms/) project where the implementation is forced to create an unstructured task for iterating the upstream sequences, which must outlive downstream calls. These libraries would like to participate in task priority escalation to boost the priority of the upstream consuming task, however today lack the API to do so.

Reviewer 3: ```suggestion
One such example is the async sequence [`merge`](https://github.com/apple/swift-async-algorithms/blob/4c3ea81f81f0a25d0470188459c6d4bf20cf2f97/Sources/AsyncAlgorithms/AsyncAlgorithms.docc/Guides/Merge.md) operation from the [swift-async-algorithms](https://github.com/apple/swift-async-algorithms/) project where the implementation is forced to create an unstructured task for iterating the upstream sequences, which must outlive downstream calls. These libraries would like to participate in task priority escalation to boost the priority of the upstream consuming task, however today they lack the API to do so.
```

struct AsyncMergeSequenceIterator: AsyncIterator {
  struct State {
    var task: Task<Void, any Error>? // unstructured upstream consumer task
    var buffer: Deque<Element>
    var upstreamContinuations: [UnsafeContinuation<Void, Error>]
    var downstreamContinuation: UnsafeContinuation<Element?, Error>?
  }
  
  let state = Mutex<State>(State())
  
  func next() async throws {
    self.state.withLock { state in
      if state.task == nil {
        state.task = Task {
          // Consume from the base iterators
          // ... 
        }
      }
    }

    if let element = self.state.withLock { $0.buffer.popFirst() } {
      return element
    } else {
      // We are handling cancellation here and need to handle task escalation here as well
      try await withTaskCancellationHandler { 
        // HERE: need to handle priority escalation and boost `state.task`
        try await withCheckedContinuation { cont in
          self.state.withLock { $0.consumerContinuation = cont }
        }
      } onCancel: {
        // trigger cancellation of tasks and fail continuations
      }
    }
  }
}
```

> The above example showcases a common pattern: often an unchecked continuation is paired with a Task used to complete it. Around the suspension on the continuation, waiting for it to be resumed, developers often install a task cancellation handler in order to potentially break out of potentially unbounded waiting for a continuation to be resumed. Around the same suspension (marked with `HERE` in the snippet above), we might want to insert a task priority escalation handler in order to priority boost the task that is used to resume the continuation. This can be important for correctness and performance of such operations, so we should find a way to offer these libraries to participate in task priority handling.

Reviewer 4: nit:

> often an unchecked continuation

there's a mix of checked & unchecked continuations in the above example. maybe drop the additional descriptor and just go with 'continuation'.

> so we should find a way to offer these libraries [a mechanism] to participate in task priority handling.

> The above example showcases a common pattern: often an unchecked continuation is paired with a Task used to complete it. Around the suspension on the continuation, waiting for it to be resumed, developers often install a task cancellation handler in order to potentially break out of potentially unbounded waiting for a continuation to be resumed. Around the same suspension (marked with `HERE` in the snippet above), we might want to insert a task priority escalation handler in order to priority boost the task that is used to resume the continuation. This can be important for correctness and performance of such operations, so we should find a way to offer these libraries to participate in task priority handling.

Author: ```suggestion
The above example showcases a common pattern: often a continuation is paired with a Task used to complete it. Around the suspension on the continuation, waiting for it to be resumed, developers often install a task cancellation handler in order to potentially break out of potentially unbounded waiting for a continuation to be resumed. Around the same suspension (marked with `HERE` in the snippet above), we might want to insert a task priority escalation handler in order to priority boost the task that is used to resume the continuation. This can be important for correctness and performance of such operations, so we should find a way to offer these libraries to participate in task priority handling.
```

Another example of libraries which may want to reach for manual task priority escalation APIs are libraries which facilitate communication across process boundaries, and would like to react to priority escalation and propagate it to a different process. Relying on the built-in priority escalation mechanisms won't work, because they are necessarily in-process, so libraries like this need to be able to participate and be notified when priority escalation happens, and also be able to efficiently cause the escalation inside the other process.

## Proposed solution

> In order to address the above use-cases, we propose to add a pair of APIs: to react to priority escalation happening within a block of code, and an API to _cause_ a priority escalation without resorting to trickery using creating new tasks whose only purpose is to escalate the priority of some other task:

Reviewer 4: ```suggestion
In order to address the above use-cases, we propose to add a pair of APIs: to react to priority escalation happening within a block of code, and an API to _cause_ a priority escalation without resorting to trickery by creating new tasks whose only purpose is to escalate the priority of some other task:
```

```swift
enum State {
  case initialized
  case task(Task<Void, Never>)
  case priority(TaskPriority)
}
let m: Mutex<State> = .init(.initialized) 

await withTaskPriorityEscalationHandler {
  await withCheckedContinuation { cc in
    let task = Task { cc.resume() }
    
    let newPriority: TaskPriority? = state.withLock { state -> TaskPriority? in
      defer { state = .task(task) }
      switch state {
      case .initialized:
          return nil
      case .task:
          preconditionFailure("unreachable")
      case .priority(let priority):
          return priority
      }
    }

>     // priority was escalated just before we have store the task in the mutex

Reviewer 4: ```suggestion
    // priority was escalated just before we stored the task in the mutex
```

    if let newPriority {
        Task.escalatePriority(task, to: newPriority)
    }
  } onPriorityEscalated: { newPriority in
    state.withLock { state in
      switch state {
      case .initialized, .priority:

>         // priority was escalated just before managed to store the task in the mutex

Reviewer 4: ```suggestion
        // priority was escalated just before we managed to store the task in the mutex
```
